File: packages/jexlang-java/src/main/java/io/github/vinaykumar0339/listeners/JexLangErrorListener.java
package io.github.vinaykumar0339.listeners;


import io.github.vinaykumar0339.eval.errors.JexLangSyntaxError;
import io.github.vinaykumar0339.eval.errors.SyntaxErrorLocation;
import org.antlr.v4.runtime.BaseErrorListener;
import org.antlr.v4.runtime.RecognitionException;
import org.antlr.v4.runtime.Recognizer;
import org.antlr.v4.runtime.Token;

import java.util.ArrayList;

public class JexLangErrorListener extends BaseErrorListener {

    private ArrayList<JexLangSyntaxError> syntaxErrors;
    @Override
    public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int line, int charPositionInLine, String msg, RecognitionException e) {

        // Extract offending text if available
        String offendingText = null;
        if (offendingSymbol instanceof Token t) {
            if (t.getText() != null) offendingText = t.getText();
        }

        String formattedMessage = formatErrorMessage(line, charPositionInLine, msg, offendingText);

        JexLangSyntaxError syntaxError = new JexLangSyntaxError(
                formattedMessage,
                new SyntaxErrorLocation(
                        line,
                        charPositionInLine,
                        offendingText
                )
        );

        syntaxErrors.add(syntaxError);
    }

    private String formatErrorMessage(int line, int charPositionInLine, String msg, String offendingTextOrNull) {
        String cleanMessage = msg
                .replaceAll("no viable alternative at input", "Unexpected")
                .replaceAll("extraneous input", "Unexpected")
                .replaceAll("mismatched input", "Unexpected")
                .replaceAll("expecting", "expected");

        if (offendingTextOrNull != null) {
            cleanMessage = cleanMessage.replaceAll(offendingTextOrNull, this.escapeString(offendingTextOrNull));
        }


        return "Syntax error at line " + line + ":" + (charPositionInLine) + ":" + cleanMessage;
    }


    private String escapeString(String str) {
        return str
                .replaceAll("\n", "\\\\n")
                .replaceAll("\r", "\\\\r")
                .replaceAll("\t", "\\\\t");
    }

    public boolean hasErrors() {
        return syntaxErrors != null && !syntaxErrors.isEmpty();
    }

    public ArrayList<JexLangSyntaxError> getSyntaxErrors() {
        return syntaxErrors;
    }

    public void clear() {
        if (syntaxErrors != null) {
            syntaxErrors.clear();
        } else {
            syntaxErrors = new ArrayList<>();
        }
    }

}

File: packages/jexlang-swift/jexlang-swift/eval/listeners/JexLangErrorListener.swift
//
//  JexLangErrorListener.swift
//  jexlang-swift
//
//  Created by Vinay Kumar on 03/10/25.
//

import Foundation

public class JexLangErrorListener: BaseErrorListener {
    private var syntaxErrors: [JexLangSyntaxError] = []
    
    public func hasErrors() -> Bool {
        return !syntaxErrors.isEmpty
    }
    
    public func getSyntaxErrors() -> [JexLangSyntaxError] {
        return self.syntaxErrors;
    }
    
    public override func syntaxError<T>(_ recognizer: Recognizer<T>, _ offendingSymbol: AnyObject?, _ line: Int, _ charPositionInLine: Int, _ msg: String, _ e: AnyObject?) where T : ATNSimulator {
        // Extract offending text if available
        var offendingText: String? = nil;
        
        if let token = offendingSymbol as? Token {
            if let text = token.getText() {
                offendingText = text
            }
        }
        
        let formattedMessage = formatErrorMessage(line: line, charPositionInLine: charPositionInLine, msg: msg, offendingText: offendingText)
        
        let syntaxError = JexLangSyntaxError(message: formattedMessage, location: SyntaxErrorLocation(line: line, column: charPositionInLine, offendingSymbol: offendingText))
        
        syntaxErrors.append(syntaxError)
    }
    
    private func formatErrorMessage(line: Int, charPositionInLine: Int, msg: String, offendingText: String?) -> String {
        var cleanMessage = msg
            .replacingOccurrences(of: "no viable alternative at input", with: "Unexpected")
            .replacingOccurrences(of: "extraneous input", with: "Unexpected")
        
        if let offendingText = offendingText {
            cleanMessage = cleanMessage.replacingOccurrences(of: offendingText, with: self.escapeString(offendingText))
        }
        
        return "Syntax error at line \(line):\(charPositionInLine): \(cleanMessage)"
    }
    
    private func escapeString(_ str: String) -> String {
        return str
            .replacingOccurrences(of: "\n", with: "\\n")
            .replacingOccurrences(of: "\r", with: "\\r")
            .replacingOccurrences(of: "\t", with: "\\t")
    }
    
    public func clear() {
        self.syntaxErrors.removeAll()
    }
}
